### ISSUE TITLE
Controller Crash: "fatal error: concurrent map writes"

### ISSUE THREAD

**Issue by wanghong230** [argoproj/argo#4262]
## Summary 

What happened/what you expected to happen?

```
fatal error: concurrent map writes

goroutine 207 [running]:
runtime.throw(0x184efcd, 0x15)
        /usr/local/go/src/runtime/panic.go:774 +0x72 fp=0xc0006094d0 sp=0xc0006094a0 pc=0x42e5d2
runtime.mapassign_faststr(0x161a4e0, 0xc000412180, 0xc0014158c0, 0x31, 0xc000a04140)
        /usr/local/go/src/runtime/map_faststr.go:211 +0x417 fp=0xc000609538 sp=0xc0006094d0 pc=0x413127
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).computeMetrics(0xc0013d80a0, 0xc0018a7b90, 0x1, 0x1, 0xc0008ab560, 0xc0008ab530, 0x0)
        /go/src/github.com/argoproj/argo/workflow/controller/operator.go:2536 +0xe7b fp=0xc0006096e8 sp=0xc000609538 pc=0x14bfdcb
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).executeTemplate(0xc0013d80a0, 0xc001415780, 0x36, 0x1a82a60, 0xc001074bb0, 0xc000fd9000, 0xc000d1fea0, 0x2, 0x4, 0x0, ...)
        /go/src/github.com/argoproj/argo/workflow/controller/operator.go:1403 +0x553 fp=0xc00060ac60 sp=0xc0006096e8 pc=0x14b0423
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).executeStepGroup(0xc0013d80a0, 0xc001074b00, 0x1, 0x4, 0xc0012bc600, 0x21, 0xc00060c248, 0x0)
        /go/src/github.com/argoproj/argo/workflow/controller/steps.go:233 +0x10cd fp=0xc00060bf70 sp=0xc00060ac60 pc=0x14c514d
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).executeSteps(0xc0013d80a0, 0xc0018cfac0, 0x1e, 0xc000fd9000, 0xc00142ae70, 0x24, 0xc000f78900, 0x1a82a60, 0xc000f5e210, 0xc00060e658, ...)
        /go/src/github.com/argoproj/argo/workflow/controller/steps.go:91 +0x46b fp=0xc00060cda8 sp=0xc00060bf70 pc=0x14c22fb
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).executeTemplate(0xc0013d80a0, 0xc0018cfac0, 0x1e, 0x1a82a60, 0xc000f5e210, 0xc000fd8fc0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/argoproj/argo/workflow/controller/operator.go:1493 +0x1808 fp=0xc00060e320 sp=0xc00060cda8 pc=0x14b16d8
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).operate(0xc0013d80a0)
        /go/src/github.com/argoproj/argo/workflow/controller/operator.go:281 +0xecb fp=0xc00060fa28 sp=0xc00060e320 pc=0x14a293b
github.com/argoproj/argo/workflow/controller.(*WorkflowController).processNextItem(0xc000143600, 0x3a22657361687000)
        /go/src/github.com/argoproj/argo/workflow/controller/controller.go:480 +0x88e fp=0xc00060fe38 sp=0xc00060fa28 pc=0x1496a7e
github.com/argoproj/argo/workflow/controller.(*WorkflowController).runWorker(0xc000143600)
        /go/src/github.com/argoproj/argo/workflow/controller/controller.go:394 +0x2b fp=0xc00060fe58 sp=0xc00060fe38 pc=0x14961cb
github.com/argoproj/argo/workflow/controller.(*WorkflowController).runWorker-fm()
        /go/src/github.com/argoproj/argo/workflow/controller/controller.go:393 +0x2a fp=0xc00060fe70 sp=0xc00060fe58 pc=0x14d82ea
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000ab0180)
        /go/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:152 +0x5e fp=0xc00060fee0 sp=0xc00060fe70 pc=0x124a25e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000ab0180, 0x3b9aca00, 0x0, 0x64222c22787a3401, 0xc0003c80c0)
        /go/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:153 +0xf8 fp=0xc00060ff90 sp=0xc00060fee0 pc=0x12497f8
k8s.io/apimachinery/pkg/util/wait.Until(0xc000ab0180, 0x3b9aca00, 0xc0003c80c0)
        /go/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:88 +0x4d fp=0xc00060ffc8 sp=0xc00060ff90 pc=0x12496ed
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc00060ffd0 sp=0xc00060ffc8 pc=0x45b741
created by github.com/argoproj/argo/workflow/controller.(*WorkflowController).Run
```

## Diagnostics

What Kubernetes provider are you using? 

EKS 1.16

What version of Argo Workflows are you running? 

```
      containers:
      - args:
        - --configmap
        - workflow-controller-configmap
        - --executor-image
        - docker.intuit.com/argoproj/argoexec:v2.8.2
        command:
        - workflow-controller
        image: docker.intuit.com/argoproj/workflow-controller:v2.8.2
        imagePullPolicy: IfNotPresent
        name: workflow-controller
        resources:
          requests:
            cpu: 100m
            memory: 100Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
```
---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
This is caused by `Metrics               map[string]prometheus.Metric`. This does not have a guard on it. 

We could replace it with `Metrics               *sync.Map`.

@wanghong230 would you be interested is submitting a PR?

**Comment by alexec** [comment]
This has been fixed in #3350 (I think).
